include/itkPhaseAnalysisSoftThresholdImageFilter.hxx

```diff
@@ -58,59 +58,40 @@ PhaseAnalysisSoftThresholdImageFilter< TInputImage, TOutputImage >
 template< typename TInputImage, typename TOutputImage >
 void
 PhaseAnalysisSoftThresholdImageFilter< TInputImage, TOutputImage >
-::BeforeThreadedGenerateData()
+::GenerateData()
 {
-  unsigned int nC = this->GetInput()->GetNumberOfComponentsPerPixel();
-
-  if ( nC < 2 )
+  // Populate and compute outputs from superclass (threaded)
+  Superclass::GenerateData();
+  auto amplitudePtr = this->GetOutputAmplitude();
+  // Compute mean/variance only once.
+  if ( this->GetApplySoftThreshold() )
     {
-    itkExceptionMacro(
-        << "Number of components of input image (" << nC
-        << ") is less than 2. PhaseAnalysis require at least 2 components.");
+    using StatisticsImageFilter = itk::StatisticsImageFilter< OutputImageType >;
+    auto statsFilter = StatisticsImageFilter::New();
+    statsFilter->SetInput(amplitudePtr);
+    statsFilter->Update();
+    this->m_MeanAmp   = statsFilter->GetMean();
+    this->m_SigmaAmp  = sqrt(statsFilter->GetVariance());
+    this->m_Threshold = this->m_MeanAmp + this->m_NumOfSigmas * this->m_SigmaAmp;
     }
 
-  // Instead of using GetNumberOfThreads, we need to split the image into the
-  // number of regions that will actually be returned by
-  // itkImageSource::SplitRequestedRegion. Sometimes this number is less than
-  // the number of threads requested.
-  OutputImageRegionType dummy;
-  unsigned int actualThreads = this->SplitRequestedRegion(
-      0, this->GetNumberOfThreads(), dummy);
-
-  m_Barrier1 = Barrier::New();
-  m_Barrier1->Initialize(actualThreads);
-  m_Barrier2 = Barrier::New();
-  m_Barrier2->Initialize(actualThreads);
+  this->GetMultiThreader()->template ParallelizeImageRegion<TOutputImage::ImageDimension>(
+    this->GetOutput()->GetRequestedRegion(),
+    [this](const OutputImageRegionType & outputRegionForThread)
+    { this->ThreadedComputeCosineOfPhase(outputRegionForThread); },
```

Comment:
You could have just called `DynamicThreadedGenerateData`. You didn't need to rename it.